hathor/consensus/poa/poa_block_producer.py

```diff
@@ -0,0 +1,173 @@
+        '_last_seen_best_block',
+    )
+
+    def __init__(
+        self,
+        *,
+        settings: HathorSettings,
+        reactor: ReactorProtocol,
+        manager: 'HathorManager',
+        poa_signer: PoaSigner,
+    ) -> None:
+        assert isinstance(settings.CONSENSUS_ALGORITHM, PoaSettings)
+        self._log = logger.new()
+        self._settings = settings
+        self._poa_settings = settings.CONSENSUS_ALGORITHM
+        self._reactor = reactor
+        self._manager = manager
+        self._poa_signer = poa_signer
+        self._signer_index = self._calculate_signer_index(self._poa_settings, self._poa_signer)
+        self._last_seen_best_block: Block | None = None
+
+        self._started_producing = False
+        self._start_producing_lc = LoopingCall(self._start_producing)
+        self._start_producing_lc.clock = self._reactor
+
+        self._schedule_block_lc = LoopingCall(self._schedule_block)
+        self._schedule_block_lc.clock = self._reactor
+
+    def start(self) -> None:
+        self._start_producing_lc.start(_WAIT_SYNC_DELAY)
+        self._schedule_block_lc.start(self._settings.AVG_TIME_BETWEEN_BLOCKS)
+
+    def stop(self) -> None:
+        if self._start_producing_lc.running:
+            self._start_producing_lc.stop()
+
+        if self._schedule_block_lc.running:
+            self._schedule_block_lc.stop()
+
+    @staticmethod
+    def _calculate_signer_index(settings: PoaSettings, poa_signer: PoaSigner) -> int:
+        """Return the signer index for the given private key."""
+        public_key = poa_signer.get_public_key()
+        public_key_bytes = get_public_key_bytes_compressed(public_key)
+        sorted_signers = sorted(settings.signers)
+        try:
+            return sorted_signers.index(public_key_bytes)
+        except ValueError:
+            raise ValueError(f'Public key "{public_key_bytes.hex()}" not in list of PoA signers')
+
+    def _start_producing(self) -> None:
+        """Start producing new blocks."""
+        if not self._manager.can_start_mining():
+            # We're syncing, so we'll try again later
+            self._log.warn('cannot start producing new blocks, node not synced')
+            return
+
+        self._log.info('started producing new blocks')
+        self._started_producing = True
+        self._start_producing_lc.stop()
+
+    def _schedule_block(self) -> None:
+        """Schedule propagation of a new block."""
+        previous_block = self._manager.tx_storage.get_best_block()
+        if not self._started_producing or previous_block == self._last_seen_best_block:
+            return
+
+        self._last_seen_best_block = previous_block
+        now = self._reactor.seconds()
+        expected_timestamp = self._expected_block_timestamp(previous_block)
+        propagation_delay = 0 if expected_timestamp < now else expected_timestamp - now
+
+        self._reactor.callLater(propagation_delay, self._produce_block, previous_block)
```

Comment:
You don't have any safety check whether there are two or more scheduled calls to `self._produce_block`. This might happen if the `propagation_delay` gets too large and this method is called again by the looping call. You can easily do it saving the `DelayedCall` returned by the `callLater()` call and set it back to none in the `_produce_block()` method.

Reply:
Done in a separate PR to make it easy to update future tests: https://github.com/HathorNetwork/hathor-core/pull/1080

Reply:
Thanks!